### PR TITLE
Matrix multiplication using Shared Memory

### LINKED ISSUE #8: Bug in matrix multiplication using shared memory
#### Description of the problem

The following code doesn't give correct results for matrix multiplication using shared memory,
```cpp
template <class data_type_matrix>
            __device__
            data_type_matrix get_element(
            data_type_matrix* mat,
            unsigned row,
            unsigned col,
            unsigned stride)
            {
                return mat[row*stride+col];
            }

            template <class data_type_matrix>
            __device__
            void set_element(
            data_type_matrix* mat,
            unsigned row,
            unsigned col,
            data_type_matrix value,
            unsigned stride)
            {
                mat[row*stride+col] = value;
            }

            template <class data_type_matrix>
            __device__
            data_type_matrix* get_sub_matrix(
            data_type_matrix* mat,
            unsigned block_row,
            unsigned block_col,
            unsigned stride)
            {
                data_type_matrix* mat_sub =
                new data_type_matrix[BLOCK_SIZE*BLOCK_SIZE];
                mat_sub = &mat[stride*BLOCK_SIZE*block_row+BLOCK_SIZE*block_col];
                return mat_sub;
            }

            template <class data_type_matrix>
            __global__
            void multiply_kernel(
            data_type_matrix* mat1,
            data_type_matrix* mat2,
            data_type_matrix* result,
            unsigned mat1_cols,
            unsigned mat2_cols,
            unsigned result_cols)
            {
                unsigned block_row = blockIdx.y;
                unsigned block_col = blockIdx.x;
                data_type_matrix* result_sub = get_sub_matrix(result, block_row,
                                                              block_col, result_cols);

                unsigned row = threadIdx.y;
                unsigned col = threadIdx.x;

                for(unsigned m = 0; m < (mat1_cols + BLOCK_SIZE - 1)/BLOCK_SIZE; m++)
                {
                    data_type_matrix* mat1_sub = get_sub_matrix(mat1, block_row,
                                                                m, mat1_cols);
                    data_type_matrix* mat2_sub = get_sub_matrix(mat2, m,
                                                                block_col, mat2_cols);

                    __shared__ data_type_matrix mat1_shared[BLOCK_SIZE][BLOCK_SIZE];
                    __shared__ data_type_matrix mat2_shared[BLOCK_SIZE][BLOCK_SIZE];

                    mat1_shared[row][col] = get_element(mat1_sub, row, col, mat1_cols);
                    mat2_shared[row][col] = get_element(mat2_sub, row, col, mat2_cols);

                    data_type_matrix cvalue = 0.0;

                    __syncthreads();

                    for(unsigned e = 0; e < BLOCK_SIZE; e++)
                    {
                        cvalue += mat1_shared[row][e] * mat2_shared[e][col];
                    }

                    __syncthreads();

                    set_element(result_sub, row, col, cvalue, result_cols);
                }
            }

            template <class data_type_matrix>
            void multiply_gpu(const MatrixGPU<data_type_matrix>& mat1,
                              const MatrixGPU<data_type_matrix>& mat2,
                              MatrixGPU<data_type_matrix>& result)
            {
                adaboost::utils::check(mat1.get_cols() == mat2.get_rows(),
                                       "Order of matrices don't match.");
                dim3 gridDim((mat2.get_cols() + BLOCK_SIZE)/BLOCK_SIZE,
                             (mat1.get_rows() + BLOCK_SIZE)/BLOCK_SIZE);
                dim3 blockDim(BLOCK_SIZE, BLOCK_SIZE);
                multiply_kernel
                <<<gridDim, blockDim>>>
                (mat1.get_data_pointer(),
                 mat2.get_data_pointer(),
                 result.get_data_pointer(),
                 mat1.get_cols(),
                 mat2.get_cols(),
                 result.get_cols());
            }
```

#### Example of the problem

```cpp
#include<gtest/gtest.h>
#include<string>
#include<adaboost/cuda/cuda_data_structures.hpp>
#include<adaboost/utils/cuda_wrappers.hpp>
#include<stdexcept>

TEST(Cuda, MatricesGPU)
{
    adaboost::utils::cuda::cuda_event_t has_happened;
    adaboost::utils::cuda::cuda_event_create(&has_happened);
    adaboost::cuda::core::MatrixGPU<float> mat_f;
    EXPECT_EQ(0, mat_f.get_cols())<<"Number of columns should be 0";
    EXPECT_EQ(0, mat_f.get_rows())<<"Number of rows should be 0.";
    adaboost::cuda::core::MatrixGPU<float> mat1(3, 3), mat2(3, 3), mat3(2, 1);
    mat1.fill(4.0);
    mat2.fill(5.0);
    mat1.copy_to_device();
    mat2.copy_to_device();
    adaboost::utils::cuda::cuda_event_record(has_happened);
    adaboost::utils::cuda::cuda_event_synchronize(has_happened);
    adaboost::cuda::core::MatrixGPU<float> result1(3, 3);
    adaboost::cuda::core::multiply_gpu(mat1, mat2, result1);
    adaboost::utils::cuda::cuda_event_record(has_happened);
    adaboost::utils::cuda::cuda_event_synchronize(has_happened);
    result1.copy_to_host();
    for(unsigned int i = 0; i < 3; i++)
    {
        for(unsigned int j = 0; j < 3; j++)
        {
            std::cout<<i<<" "<<j<<" "<<result1.at(i, j)<<std::endl;
            EXPECT_EQ(60.0, result1.at(i, j));
        }
    }
    mat3.set(0, 0, 6.0);
    mat3.set(1, 0, 6.0);
    EXPECT_THROW({
        try
        {
            adaboost::cuda::core::multiply_gpu(mat1, mat3, result1);
        }
        catch(const std::logic_error& e)
        {
            EXPECT_STREQ("Order of matrices don't match.", e.what());
            throw;
        }
    }, std::logic_error);
}

int main(int ac, char* av[])
{
    testing::InitGoogleTest(&ac, av);
    return RUN_ALL_TESTS();
}
```

### PR DESCRIPTION
**Pull request by Tanvi141**
#### References to other Issues or PRs or Relevant literature

Fixes #8 

#### Brief description of what is fixed or changed
Made changes to `multiply_gpu` method to implement logic of tiling using shared memory. Corner conditions included in case matrix size does not exactly match size of tile.

### CODE REVIEW

File: adaboost/tests/test_cuda_data_structures.hpp [hunk 1]
```diff
@@ -77,3 +77,45 @@ TEST(Cuda, MatrixGPU)
...
+
+TEST(Cuda, MatricesGPU)
+{
+    adaboost::utils::cuda::cuda_event_t has_happened;
+    adaboost::utils::cuda::cuda_event_create(&has_happened);
+    adaboost::cuda::core::MatrixGPU<float> mat_f;
+    EXPECT_EQ(0, mat_f.get_cols())<<"Number of columns should be 0";
+    EXPECT_EQ(0, mat_f.get_rows())<<"Number of rows should be 0.";
+    adaboost::cuda::core::MatrixGPU<float> mat1(3, 3), mat2(3, 3), mat3(2, 1);
+    adaboost::cuda::core::fill(float(4.0),mat1,0,0);
```
Comment (czgdp1807) [inline adaboost/tests/test_cuda_data_structures.hpp]:
```suggestion
    adaboost::cuda::core::fill(float(4.0), mat1, 0, 0);
```
Add spaces after `,` in other places as well.

**Comment by czgdp1807** [comment]
Could you test your code using [this](https://www.mathsisfun.com/algebra/images/matrix-multiply-ex1b.gif) example?

File: adaboost/tests/test_cuda_data_structures.hpp [hunk 2]
```diff
@@ -77,3 +77,45 @@ TEST(Cuda, MatrixGPU)
...
+    adaboost::cuda::core::MatrixGPU<float> result1(3, 3);
+    adaboost::cuda::core::multiply_gpu(mat1, mat2, result1);
+    adaboost::utils::cuda::cuda_event_record(has_happened);
+    adaboost::utils::cuda::cuda_event_synchronize(has_happened);
+    result1.copy_to_host();
+    for(unsigned int i = 0; i < 3; i++)
+    {
+        for(unsigned int j = 0; j < 3; j++)
+        {
+            std::cout<<i<<" "<<j<<" "<<result1.at(i, j)<<std::endl;
```
Comment (czgdp1807) [inline adaboost/tests/test_cuda_data_structures.hpp]:
```suggestion
```

File: adaboost/tests/test_cuda_data_structures.hpp (hunk 1 above)
Comment (Tanvi141) [reply adaboost/tests/test_cuda_data_structures.hpp]: ok, will update in other places as well

**Comment by Tanvi141** [comment]
Tested the particular example, working properly.
```
[==========] Running 3 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 3 tests from Cuda
[ RUN      ] Cuda.VectorGPU
[       OK ] Cuda.VectorGPU (62 ms)
[ RUN      ] Cuda.MatrixGPU
[       OK ] Cuda.MatrixGPU (39 ms)
[ RUN      ] Cuda.MatricesGPU
83.000000 63.000000 37.000000 75.000000 [       OK ] Cuda.MatricesGPU (312 ms)
[----------] 3 tests from Cuda (413 ms total)

[----------] Global test environment tear-down
[==========] 3 tests from 1 test case ran. (413 ms total)
[  PASSED  ] 3 tests.
 ```
Also updated required changed suggested above

File: adaboost/tests/test_cuda_data_structures.hpp [hunk 3]
```diff
@@ -77,3 +77,74 @@ TEST(Cuda, MatrixGPU)
...
+    m2.copy_to_device();
+    adaboost::utils::cuda::cuda_event_record(has_happened);
+    adaboost::utils::cuda::cuda_event_synchronize(has_happened);
+    adaboost::cuda::core::MatrixGPU<float> r1(1, 4);
+    adaboost::cuda::core::multiply_gpu(m1, m2, r1);
+    adaboost::utils::cuda::cuda_event_record(has_happened);
+    adaboost::utils::cuda::cuda_event_synchronize(has_happened);
+    r1.copy_to_host();
+    for(unsigned int i = 0; i < 4; i++) printf("%f ",r1.at(0,i));
+}
```
Comment (czgdp1807) [inline adaboost/tests/test_cuda_data_structures.hpp]: Can you use, `EXPECT_EQ`  to check the correctness?

File: adaboost/tests/test_cuda_data_structures.hpp [hunk 4]
```diff
@@ -77,3 +77,74 @@ TEST(Cuda, MatrixGPU)
...
+        catch(const std::logic_error& e)
+        {
+            EXPECT_STREQ("Order of matrices don't match.", e.what());
+            throw;
+        }
+    }, std::logic_error);
+
+    //Differently sized matrices
+    adaboost::cuda::core::MatrixGPU<float> m1(1, 3), m2(3, 4);
+    m1.set(0,0,3);
```
Comment (czgdp1807) [inline adaboost/tests/test_cuda_data_structures.hpp]:
```suggestion
    m1.set(0, 0, 3);
```
Space after `,`.

Comment (Tanvi141) [reply adaboost/tests/test_cuda_data_structures.hpp]: ok, will change this as well

File: adaboost/tests/test_cuda_data_structures.hpp (hunk 3 above)
Comment (Tanvi141) [reply adaboost/tests/test_cuda_data_structures.hpp]: Yes will do

**Comment by Tanvi141** [comment]
Hi, made required changes. Please see test_cuda reports;

```
[==========] Running 3 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 3 tests from Cuda
[ RUN      ] Cuda.VectorGPU
[       OK ] Cuda.VectorGPU (252 ms)
[ RUN      ] Cuda.MatrixGPU
[       OK ] Cuda.MatrixGPU (40 ms)
[ RUN      ] Cuda.MatricesGPU
[       OK ] Cuda.MatricesGPU (342 ms)
[----------] 3 tests from Cuda (634 ms total)

[----------] Global test environment tear-down
[==========] 3 tests from 1 test case ran. (634 ms total)
[  PASSED  ] 3 tests.
```

File: adaboost/tests/test_cuda_data_structures.hpp [hunk 5]
```diff
@@ -77,3 +77,77 @@ TEST(Cuda, MatrixGPU)
...
+    adaboost::utils::cuda::cuda_event_synchronize(has_happened);
+    adaboost::cuda::core::MatrixGPU<float> r1(1, 4);
+    adaboost::cuda::core::multiply_gpu(m1, m2, r1);
+    adaboost::utils::cuda::cuda_event_record(has_happened);
+    adaboost::utils::cuda::cuda_event_synchronize(has_happened);
+    r1.copy_to_host();
+    EXPECT_EQ(83.0, r1.at(0, 0));
+	EXPECT_EQ(63.0, r1.at(0, 1));
+	EXPECT_EQ(37.0, r1.at(0, 2));
+	EXPECT_EQ(75.0, r1.at(0, 3));
```
Comment (czgdp1807) [inline adaboost/tests/test_cuda_data_structures.hpp]:
```suggestion
    EXPECT_EQ(83.0, r1.at(0, 0));
    EXPECT_EQ(63.0, r1.at(0, 1));
    EXPECT_EQ(37.0, r1.at(0, 2));
    EXPECT_EQ(75.0, r1.at(0, 3));
```
